src/Microsoft.AspNetCore.Routing/Matchers/DfaMatcherBuilder.cs

```diff
@@ -4,32 +4,37 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
-using Microsoft.AspNetCore.Routing.EndpointConstraints;
 using Microsoft.AspNetCore.Routing.Patterns;
-using Microsoft.AspNetCore.Routing.Template;
-using Microsoft.Extensions.Options;
 
 namespace Microsoft.AspNetCore.Routing.Matchers
 {
     internal class DfaMatcherBuilder : MatcherBuilder
     {
-        private readonly List<MatcherBuilderEntry> _entries = new List<MatcherBuilderEntry>();
-        private readonly IInlineConstraintResolver _constraintResolver = new DefaultInlineConstraintResolver(Options.Create(new RouteOptions()));
+        private readonly List<MatcherEndpoint> _endpoints = new List<MatcherEndpoint>();
 
         private readonly MatchProcessorFactory _matchProcessorFactory;
-        private readonly EndpointSelector _endpointSelector;
+        private readonly EndpointSelector _selector;
+        private readonly MatcherPolicy[] _policies;
+        private readonly INodeBuilderPolicy[] _nodeBuilders;
+        private readonly MatcherEndpointComparer _comparer;
 
         public DfaMatcherBuilder(
             MatchProcessorFactory matchProcessorFactory,
-            EndpointSelector endpointSelector)
+            EndpointSelector selector,
+            IEnumerable<MatcherPolicy> policies)
         {
-            _matchProcessorFactory = matchProcessorFactory ?? throw new ArgumentNullException(nameof(matchProcessorFactory));
-            _endpointSelector = endpointSelector ?? throw new ArgumentNullException(nameof(endpointSelector));
+            _matchProcessorFactory = matchProcessorFactory;
+            _selector = selector;
+            _policies = policies.OrderBy(p => p.Order).ToArray();
```

Comment:
OK so this is where you see `MatcherPolicy` for the first time. The idea with this is that it's a feature-collection or capability-based thing. Right now matcher policies can:
- order endpoints
- add nodes to the graph
- both
- neither

If we add more capabilites in the future, we don't need to plumb them through everything.